istioctl/cmd/vm-bootstrap.go

```diff
@@ -0,0 +1,615 @@
+	sshIgnoreHostKeys    bool
+	sshUser              string
+	startIstio           bool
+	workloadNamespace    string
+
+	workloadKind = collections.IstioNetworkingV1Alpha3Workloadentries.Resource().GroupVersionKind()
+)
+
+type workloadEntryAddressKeys struct {
+	CaCert           []byte
+	Cert             []byte
+	Key              []byte
+	ServiceName      string
+	ServiceNamespace string
+}
+
+type citadelDescribedCert struct {
+	PemEncodedCa []byte
+	Ca           *x509.Certificate
+	Key          crypto.PrivateKey
+}
+
+type remoteResponse struct {
+	typ     uint8
+	message string
+}
+
+func fetchSingleWorkloadEntry(workloadName string, client model.ConfigStore) ([]model.Config, string, error) {
+	workloadSplit := strings.Split(workloadName, "/")
+	if len(workloadSplit) != 2 {
+		return nil, "", fmt.Errorf("workload name: %s is not in the format: workloadName/Namespace", workloadName)
+	}
+
+	we := client.Get(workloadKind, workloadSplit[0], workloadSplit[1])
+	if we == nil {
+		return nil, "", fmt.Errorf("workload entry: %s in namespace: %s was not found", workloadSplit[0], workloadSplit[1])
+	}
+
+	return []model.Config{*we}, workloadSplit[1], nil
+}
+
+func fetchAllWorkloadEntries(client model.ConfigStore) ([]model.Config, string, error) {
+	list, err := client.List(workloadKind, workloadNamespace)
+	return list, workloadNamespace, err
+}
+
+func getCitadelCertificate(kubeClient kubernetes.Interface) (citadelDescribedCert, error) {
+	secret, err := kubeClient.CoreV1().Secrets(istioSystemNamespace).Get(context.TODO(), "istio-ca-secret", metav1.GetOptions{})
+	if err != nil {
+		return citadelDescribedCert{}, err
+	}
+
+	key, err := util.ParsePemEncodedKey(secret.Data["ca-key.pem"])
+	if err != nil {
+		return citadelDescribedCert{}, err
+	}
+	certContents := secret.Data["ca-cert.pem"]
+	cert, err := util.ParsePemEncodedCertificate(certContents)
+	if err != nil {
+		return citadelDescribedCert{}, err
+	}
+
+	return citadelDescribedCert{
+		PemEncodedCa: certContents,
+		Ca:           cert,
+		Key:          key,
+	}, nil
+}
+
+func extractOrgName(citadelCert *x509.Certificate) string {
+	return citadelCert.Subject.Organization[0]
+}
+
+func getCertificatesForEachAddress(
+	workloadEntries []model.Config,
+	namespace string,
+	root citadelDescribedCert) (map[string]workloadEntryAddressKeys, error) {
+	seenIps := make(map[string]workloadEntryAddressKeys)
+
+	if organization == "" {
+		organization = extractOrgName(root.Ca)
+	}
+
+	for _, entryCfg := range workloadEntries {
+		wle := entryCfg.Spec.(*networking.WorkloadEntry)
+		// Only generate one certificate per address.
+		if _, ok := seenIps[wle.Address]; ok {
+			continue
+		}
+
+		signerOpts := util.CertOptions{
+			Host:         wle.Address,
+			NotBefore:    time.Now(),
+			TTL:          certDuration,
+			SignerCert:   root.Ca,
+			SignerPriv:   root.Key,
+			Org:          organization,
+			IsCA:         false,
+			IsSelfSigned: false,
+			IsClient:     true,
+			RSAKeySize:   4096,
```

Comment:
I think 2048 is the default in the rest of the code

Reply:
Will change.